### PR TITLE
DOT-4767: When using Standalone mode, the user should be directed to the payment page with the language of the store

### PR DESCRIPTION
**Pull request by olyans**

### CODE REVIEW

File: includes/class-wc-youcanpay-api.php [hunk 1]
```diff
@@ -4,247 +4,264 @@
...
+        }
+
+        return self::$is_test_mode;
+    }
+
+    /**
+     * Send the request to YouCan Pay's API
+     *
+     * @param WC_Order|WC_Order_Refund $order
+     * @param array $post_data
```
Comment (AmineAlyate) [inline includes/class-wc-youcanpay-api.php:132]: the phpdoc should contain the exceptions also 🙏

File: includes/class-wc-youcanpay-api.php [hunk 2]
```diff
@@ -4,247 +4,264 @@
...
+            WC_YouCanPay_Logger::alert('throwable at request exists into wc youcan pay api', [
+                'exception.message' => $e->getMessage(),
+            ]);
+        }
+
+        return json_decode(json_encode($response));
+    }
+
+    /**
+     * @return Transaction|null
```
Comment (AmineAlyate) [inline includes/class-wc-youcanpay-api.php:190]: @param

File: includes/class-wc-youcanpay-api.php [hunk 3]
```diff
@@ -4,247 +4,264 @@
...
+            $ip = $_SERVER['HTTP_CLIENT_IP'];
+        } elseif (!empty($_SERVER['HTTP_X_FORWARDED_FOR'])) {
+            //to check ip is pass from proxy
+            $ip = $_SERVER['HTTP_X_FORWARDED_FOR'];
+        } else {
+            $ip = $_SERVER['REMOTE_ADDR'];
+        }
+
+        return $ip;
+    }
```
Comment (AmineAlyate) [inline includes/class-wc-youcanpay-api.php:265]:
```
    public static function get_the_user_ip()
    {
        if (!empty($_SERVER['HTTP_CLIENT_IP'])) {
            //check ip from share internet
            return  $_SERVER['HTTP_CLIENT_IP'];
        }
        
        if (!empty($_SERVER['HTTP_X_FORWARDED_FOR'])) {
            //to check ip is pass from proxy
            return $_SERVER['HTTP_X_FORWARDED_FOR'];
        }

        return $_SERVER['REMOTE_ADDR'];
    }
```

Comment (AmineAlyate) [reply includes/class-wc-youcanpay-api.php:265]: we can refactor this 🙏

File: includes/payment-methods/class-wc-gateway-youcanpay-standalone.php [hunk 4]
```diff
@@ -1,238 +1,247 @@
...
+            'redirect' => ['return_url' => $return_url],
+        ];
+
+        return WC_YouCanPay_API::request($order, $post_data);
+    }
+
+    /**
+     * Process the payment
+     *
+     * @param int $order_id Reference.
```
Comment (AmineAlyate) [inline includes/payment-methods/class-wc-gateway-youcanpay-standalone.php:184]: for the the phpdoc should contain exceptions same for other phpdocs 🙏

File: includes/class-wc-youcanpay-api.php (hunk 1 above)
Comment (olyans) [reply includes/class-wc-youcanpay-api.php:132]: why? always we have only array as response

File: includes/payment-methods/class-wc-gateway-youcanpay-standalone.php (hunk 4 above)
Comment (olyans) [reply includes/payment-methods/class-wc-gateway-youcanpay-standalone.php:184]: @AmineAlyate Why? can you please explain more? we always have only array response

File: includes/class-wc-youcanpay-api.php (hunk 2 above)
Comment (olyans) [reply includes/class-wc-youcanpay-api.php:191]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/3583c00247402892d45f5c0af08d6827e0a5e5b5

File: includes/class-wc-youcanpay-api.php (hunk 3 above)
Comment (olyans) [reply includes/class-wc-youcanpay-api.php]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/3583c00247402892d45f5c0af08d6827e0a5e5b5

File: includes/class-wc-youcanpay-api.php [hunk 5]
```diff
@@ -4,247 +4,264 @@
...
-			$ip = $_SERVER['HTTP_X_FORWARDED_FOR'];
-		} else {
-			$ip = $_SERVER['REMOTE_ADDR'];
-		}
-
-		return $ip;
-	}
+class WC_YouCanPay_API
+{
+
```
[the remark this answers is not in the document]
Comment (olyans) [reply includes/class-wc-youcanpay-api.php]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/3583c00247402892d45f5c0af08d6827e0a5e5b5

File: includes/class-wc-youcanpay-api.php [hunk 6]
```diff
@@ -4,247 +4,268 @@
...
+                $post_data['redirect']['return_url'],
+                $post_data['redirect']['return_url']
+            );
+
+            if (is_wp_error($token) || empty($token)) {
+                WC_YouCanPay_Logger::info('there was a problem connecting to the YouCan Pay API endpoint', [
+                    'order_id' => $order->get_id(),
+                ]);
+
+                throw new WC_YouCanPay_Exception(
```
Comment (AmineAlyate) [inline includes/class-wc-youcanpay-api.php:167]: Unhandled exception

File: includes/class-wc-youcanpay-api.php (hunk 1 above)
Comment (olyans) [reply includes/class-wc-youcanpay-api.php:131]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/8f619601e2f3498862395a06dff350c1928dc177

File: includes/payment-methods/class-wc-gateway-youcanpay-standalone.php (hunk 4 above)
Comment (olyans) [reply includes/payment-methods/class-wc-gateway-youcanpay-standalone.php:185]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/8f619601e2f3498862395a06dff350c1928dc177

File: includes/class-wc-youcanpay-api.php (hunk 6 above)
Comment (olyans) [reply includes/class-wc-youcanpay-api.php:164]: https://github.com/NextmediaMa/youcan-pay-for-woocommerce/pull/12/commits/8f619601e2f3498862395a06dff350c1928dc177

**Review by AmineAlyate — APPROVED** [review]
LGTM :heavy_check_mark:

**Pull request merged**